# Lab-Assignment 9

- V.VISHNU DUT
- 16BCE1103


Datasets:<br>

Wine dataset and Wheat seed dataset<br>

Wine - non linear<br>
WheatSeed - Linear<br>

We first classify them and get their results and then apply dimensionality reduction and then apply classification algorithm on them, and then get their results.<br>

### Wheat seed

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
col_Names=["area", "perimeter", "compactness", "lengthOfKernel", "widthOfKernel", "asymmetryCoefficient", "lengthOfKernelGroove", "TypeOfWheatSeed"]
data = pd.read_csv("seeds_dataset.csv",names=col_Names, header = None)

In [4]:
data.head()

area  perimeter  compactness  lengthOfKernel  widthOfKernel  \
0  15.26      14.84       0.8710           5.763          3.312   
1  14.88      14.57       0.8811           5.554          3.333   
2  14.29      14.09       0.9050           5.291          3.337   
3  13.84      13.94       0.8955           5.324          3.379   
4  16.14      14.99       0.9034           5.658          3.562   

   asymmetryCoefficient  lengthOfKernelGroove  TypeOfWheatSeed  
0                 2.221                 5.220                1  
1                 1.018                 4.956                1  
2                 2.699                 4.825                1  
3                 2.259                 4.805                1  
4                 1.355                 5.175                1

In [5]:
data.isnull().any()

area                    False
perimeter               False
compactness             False
lengthOfKernel          False
widthOfKernel           False
asymmetryCoefficient    False
lengthOfKernelGroove    False
TypeOfWheatSeed         False
dtype: bool

In [6]:
from sklearn.preprocessing import StandardScaler

In [7]:
scaler = StandardScaler()

In [8]:
X = data[["area","perimeter","compactness","lengthOfKernel","widthOfKernel","asymmetryCoefficient","lengthOfKernelGroove"]].values

In [13]:
Y = data[["TypeOfWheatSeed"]].values

In [18]:
from sklearn.model_selection import train_test_split  
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=0)  

In [19]:
from sklearn.linear_model import LinearRegression  
regressor = LinearRegression()  
regressor.fit(X_train, y_train)  

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [20]:
print(regressor.intercept_)

[49.32930855]


In [21]:
print(regressor.coef_) 

[[  1.29678468  -2.87904628 -29.27178945  -2.44906132   0.45732162
    0.12881309   2.34067356]]


In [22]:
y_pred = regressor.predict(X_test)

In [23]:
from sklearn import metrics  
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred))  
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))  
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))  

Mean Absolute Error: 0.26781816440579814
Mean Squared Error: 0.10513096161958353
Root Mean Squared Error: 0.3242390501151635


In [30]:
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression()
clf.fit(X_train, y_train)

/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py:578: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [31]:
Y_pred = clf.predict(X_test)

In [32]:
from sklearn.metrics import accuracy_score

In [33]:
accuracy_score(Y_pred, y_test)

0.9285714285714286

In [35]:
clf2 = LogisticRegression(max_iter=1000)
clf.fit(X_train, y_train)

/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py:578: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [36]:
Y_pred = clf.predict(X_test)

In [38]:
accuracy_score(Y_pred, y_test)

0.9285714285714286

### Dimensionality reduction PCA and Kernel PCA

In [50]:
# Load libraries
from sklearn.decomposition import PCA, KernelPCA
from sklearn.datasets import make_circles

In [121]:
kpca = KernelPCA(kernel="rbf", gamma=15)
X_kpca = kpca.fit_transform(X)

pca = PCA(n_components='mle', svd_solver='full')
X_pca = pca.fit_transform(X)

In [122]:
print('Original number of features:', X.shape[1])
print('Reduced number of features KPCA:', X_kpca.shape[1])
print('Reduced number of features PCA:', X_pca.shape[1])

Original number of features: 7
Reduced number of features KPCA: 210
Reduced number of features PCA: 6


In [123]:
from sklearn.model_selection import train_test_split  
X_train, X_test, y_train, y_test = train_test_split(X_kpca, Y, test_size=0.2, random_state=0)  

In [124]:
clf2 = LogisticRegression(max_iter=1000)
clf.fit(X_train, y_train)

/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py:578: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [125]:
Y_pred = clf.predict(X_test)

In [126]:
accuracy_score(Y_pred, y_test)

0.6904761904761905

In [127]:
from sklearn.model_selection import train_test_split  
X_train, X_test, y_train, y_test = train_test_split(X_pca, Y, test_size=0.2, random_state=0)  

In [128]:
clf2 = LogisticRegression(max_iter=1000)
clf.fit(X_train, y_train)

/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py:578: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [129]:
Y_pred = clf.predict(X_test)

In [130]:
accuracy_score(Y_pred, y_test)

0.9047619047619048

In [137]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA

In [138]:
clf = LDA()

In [140]:
clf.fit(X, Y)

/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py:578: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LinearDiscriminantAnalysis(n_components=None, priors=None, shrinkage=None,
              solver='svd', store_covariance=False, tol=0.0001)

In [141]:
from sklearn.model_selection import train_test_split  
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=0)  

In [143]:
clf2 = LogisticRegression(max_iter=1000)
clf.fit(X_train, y_train)
Y_pred = clf.predict(X_test)
accuracy_score(Y_pred, y_test)

/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py:578: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


1.0

### Wine dataset

In [149]:
#We will be doing Simple Linear Regression from scratch with one attribute and not multiple
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import seaborn as sb
from sklearn import preprocessing
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [150]:
data = pd.read_csv("wineQualityReds.csv")
data.tail()

fixed.acidity  volatile.acidity  citric.acid  residual.sugar  chlorides  \
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free.sulfur.dioxide  total.sulfur.dioxide  density    pH  sulphates  \
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
1594     10.5        5  
1595     11.2        6  
1596     11.0        6  
1597     10.2        5  
1598     11.0        6

In [151]:
#data set framing
X = data[["fixed.acidity","volatile.acidity","citric.acid","residual.sugar","chlorides","free.sulfur.dioxide","density","pH","sulphates","alcohol","quality"
]]
y = data["total.sulfur.dioxide"]

X_train, X_test, y_train, y_test = train_test_split(X, y)

In [152]:
regressor = LinearRegression()  
regressor.fit(X_train, y_train)  

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [153]:
coeff_df = pd.DataFrame(regressor.coef_, X.columns, columns=['Coefficient'])  
coeff_df  

Coefficient
fixed.acidity          -8.485644
volatile.acidity       31.097172
citric.acid            51.330384
residual.sugar          0.214701
chlorides             -90.204921
free.sulfur.dioxide     1.923942
density              3052.535744
pH                    -50.852843
sulphates               9.164450
alcohol                -1.556867
quality                -4.113030

In [154]:
y_pred = regressor.predict(X_test)
df = pd.DataFrame({'Actual': y_test, 'Predicted': y_pred})  
df  

Actual   Predicted
336     16.0   19.330052
82      73.0   75.497612
235     38.0   37.493413
607     52.0   68.819998
66      39.0   33.728249
1246    48.0   54.755397
48      12.0   36.596074
1531    28.0   43.944190
1021     8.0    7.447259
939     11.0   18.225564
166    101.0   54.862939
1098    16.0    9.902397
1286    31.0   39.593095
752     66.0   63.447549
1555    24.0   36.423453
594     38.0   59.418472
701     25.0   28.469457
1384    84.0   69.800687
1306    91.0   61.248755
1325    34.0   48.948733
993     68.0   76.863455
1324    34.0   48.948733
537     13.0   32.607912
1526    38.0   50.298491
1064    10.0   14.741198
1595    51.0   88.111080
1513    65.0   45.520873
1561   131.0   87.757572
1461    13.0   34.208826
664     54.0   48.038298
...      ...         ...
1580    26.0   27.050889
1015    27.0   39.748566
182     42.0   48.410790
1125    27.0   27.092192
1435    95.0  132.043973
386     48.0   64.339721
1187    39.0   20.459237
1420    19.0   17.486901
297     29.0   25.954775
846     31.0   49.634510
815     38.0   43.256737
1147    14.0   17.531492
1587    28.0   47.629091
683     15.0   33.701577
1079   278.0  104.478628
1502    60.0   52.652821
1071    92.0  100.662057
1395    12.0   16.653327
778     34.0   31.827588
535     28.0    3.153583
688      9.0   21.442490
375     15.0   34.523217
671     13.0   39.195501
320     59.0   51.141210
264     27.0   29.438706
127     12.0   22.317761
981     28.0   48.328948
395     38.0   26.123192
909     11.0   25.492488
128     16.0   11.821643

[400 rows x 2 columns]

In [155]:
from sklearn import metrics  
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred))  
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))  
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))  

Mean Absolute Error: 15.893100141029436
Mean Squared Error: 560.1825876096477
Root Mean Squared Error: 23.668176685364838


### Now applying PCA , kernelPCA

In [156]:
# Load libraries
from sklearn.decomposition import PCA, KernelPCA
from sklearn.datasets import make_circles
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA

In [157]:
kpca = KernelPCA(kernel="rbf", gamma=15)
X_kpca = kpca.fit_transform(X)

pca = PCA(n_components='mle', svd_solver='full')
X_pca = pca.fit_transform(X)

In [158]:
print('Original number of features:', X.shape[1])
print('Reduced number of features KPCA:', X_kpca.shape[1])
print('Reduced number of features PCA:', X_pca.shape[1])

Original number of features: 11
Reduced number of features KPCA: 1495
Reduced number of features PCA: 10


In [159]:
from sklearn.model_selection import train_test_split  
X_train, X_test, y_train, y_test = train_test_split(X_kpca, Y, test_size=0.2, random_state=0)  

In [160]:
regressor = LinearRegression()  
regressor.fit(X_train, y_train)  

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [161]:
y_pred = regressor.predict(X_test)
df = pd.DataFrame({'Actual': y_test, 'Predicted': y_pred})  
df  

Actual     Predicted
0    0.593954 -2.411711e+11
1   -0.987312  1.108097e+12
2   -0.591995 -6.412048e-01
3   -0.987312  2.404462e+11
4   -1.230584 -9.299545e-01
5    0.928453  9.379654e-01
6   -0.804858  2.940471e+11
7    0.259456  2.170715e-01
8   -0.835267  3.211203e+12
9    0.928453  1.106855e+12
10  -0.865676 -5.894671e+12
11  -0.561586 -5.589676e-01
12  -0.987312 -9.530945e-01
13   1.506223  3.958935e+12
14  -0.470360 -7.572021e+12
15   0.411500  1.820227e+12
16  -1.108948 -9.164124e-01
17  -0.804858  1.717670e+12
18  -0.622404  5.360836e+12
19  -0.531178  1.031145e+13
20  -1.078539 -1.694743e+12
21  -0.652813 -9.844002e+11
22  -0.470360 -6.030197e-01
23   2.023176 -3.280201e+12
24   0.077002  4.206639e+12
25   2.631355  4.629194e+11
26   1.749495  5.935104e+12
27  -0.287906  1.508516e+12
28   2.205630  8.302017e+12
29  -0.075043  3.623595e+12
..        ...           ...
290  0.137820  1.123962e-01
291 -0.926494 -5.540977e+10
292 -0.561586 -2.645968e+01
293 -1.048130 -5.641118e+12
294  0.229047 -1.187560e+13
295 -0.774449 -8.581390e-01
296  0.685181 -3.724584e+11
297 -0.105452  3.387915e+12
298  0.472318 -3.821925e+12
299  0.928453 -2.112385e+11
300 -0.409542 -4.167023e-01
301 -1.048130 -4.419251e+12
302 -0.652813 -4.811096e-01
303  0.411500 -2.514300e+12
304 -0.835267  7.679669e+11
305  0.381091  5.229362e+12
306 -0.713631  1.276937e+12
307 -0.804858  1.145896e+12
308  1.019680 -5.197279e+12
309 -0.196679  9.231567e-04
310  1.810313  1.801125e+00
311 -0.196679 -1.283923e+03
312 -0.287906 -3.484879e-01
313 -0.987312 -2.590546e+12
314  0.624363  1.304407e+13
315 -0.622404  2.527383e+11
316 -1.200175 -4.329285e+12
317 -0.865676 -1.428596e+12
318 -0.987312 -9.369454e+11
319 -1.139357 -5.321851e+12

[320 rows x 2 columns]

In [162]:
from sklearn import metrics  
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred))  
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))  
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))  

Mean Absolute Error: 2584015151647.59
Mean Squared Error: 1.4246020099333066e+25
Root Mean Squared Error: 3774390030101.959


In [163]:
from sklearn.model_selection import train_test_split  
X_train, X_test, y_train, y_test = train_test_split(X_pca, Y, test_size=0.2, random_state=0)  

In [164]:
regressor = LinearRegression()  
regressor.fit(X_train, y_train)  

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [165]:
y_pred = regressor.predict(X_test)
df = pd.DataFrame({'Actual': y_test, 'Predicted': y_pred})  
df  

Actual  Predicted
0    0.593954   0.290676
1   -0.987312  -0.650391
2   -0.591995  -0.571807
3   -0.987312  -0.571446
4   -1.230584  -0.944468
5    0.928453  -0.203910
6   -0.804858  -0.303015
7    0.259456   1.036284
8   -0.835267  -0.463531
9    0.928453   0.404223
10  -0.865676  -0.313713
11  -0.561586  -0.498554
12  -0.987312  -0.804470
13   1.506223   1.476519
14  -0.470360   0.330161
15   0.411500   1.191559
16  -1.108948  -1.071960
17  -0.804858  -0.385060
18  -0.622404  -0.495913
19  -0.531178  -0.379800
20  -1.078539  -1.039068
21  -0.652813  -0.498470
22  -0.470360  -0.625588
23   2.023176   0.321109
24   0.077002   0.165872
25   2.631355   1.221947
26   1.749495  -0.241538
27  -0.287906   0.122317
28   2.205630   1.135510
29  -0.075043  -0.764958
..        ...        ...
290  0.137820   0.444521
291 -0.926494  -0.899741
292 -0.561586  -1.463672
293 -1.048130  -0.983332
294  0.229047   0.097240
295 -0.774449  -0.230139
296  0.685181   0.303135
297 -0.105452   0.017481
298  0.472318  -0.191399
299  0.928453   0.172745
300 -0.409542  -0.506155
301 -1.048130  -0.903656
302 -0.652813  -0.227237
303  0.411500   0.783856
304 -0.835267  -0.586179
305  0.381091   1.325862
306 -0.713631  -0.579530
307 -0.804858  -0.482637
308  1.019680   0.441848
309 -0.196679   0.117203
310  1.810313   1.754815
311 -0.196679   0.733378
312 -0.287906  -0.753661
313 -0.987312  -0.856852
314  0.624363   0.842676
315 -0.622404  -0.788178
316 -1.200175  -0.834473
317 -0.865676  -0.370189
318 -0.987312  -0.765268
319 -1.139357  -1.003652

[320 rows x 2 columns]

In [166]:
from sklearn import metrics  
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred))  
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))  
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))  

Mean Absolute Error: 0.4408379007098189
Mean Squared Error: 0.36037896897724037
Root Mean Squared Error: 0.6003157244127796


In [177]:
X = data[["fixed.acidity","volatile.acidity","citric.acid","residual.sugar","chlorides","free.sulfur.dioxide","density","pH","sulphates","alcohol","quality"
]].values
y = data[["quality"]].values

In [179]:
clf = LDA()
clf.fit(X, y) 

/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py:578: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/anaconda3/lib/python3.6/site-packages/sklearn/discriminant_analysis.py:388: UserWarning: Variables are collinear.
  warnings.warn("Variables are collinear.")


LinearDiscriminantAnalysis(n_components=None, priors=None, shrinkage=None,
              solver='svd', store_covariance=False, tol=0.0001)

In [180]:
from sklearn.model_selection import train_test_split  
X_train, X_test, y_train, y_test = train_test_split(X_pca, Y, test_size=0.2, random_state=0)  

In [181]:
regressor = LinearRegression()  
regressor.fit(X_train, y_train)  

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [182]:
y_pred = regressor.predict(X_test)
df = pd.DataFrame({'Actual': y_test, 'Predicted': y_pred})  
df  

Actual  Predicted
0    0.593954   0.290676
1   -0.987312  -0.650391
2   -0.591995  -0.571807
3   -0.987312  -0.571446
4   -1.230584  -0.944468
5    0.928453  -0.203910
6   -0.804858  -0.303015
7    0.259456   1.036284
8   -0.835267  -0.463531
9    0.928453   0.404223
10  -0.865676  -0.313713
11  -0.561586  -0.498554
12  -0.987312  -0.804470
13   1.506223   1.476519
14  -0.470360   0.330161
15   0.411500   1.191559
16  -1.108948  -1.071960
17  -0.804858  -0.385060
18  -0.622404  -0.495913
19  -0.531178  -0.379800
20  -1.078539  -1.039068
21  -0.652813  -0.498470
22  -0.470360  -0.625588
23   2.023176   0.321109
24   0.077002   0.165872
25   2.631355   1.221947
26   1.749495  -0.241538
27  -0.287906   0.122317
28   2.205630   1.135510
29  -0.075043  -0.764958
..        ...        ...
290  0.137820   0.444521
291 -0.926494  -0.899741
292 -0.561586  -1.463672
293 -1.048130  -0.983332
294  0.229047   0.097240
295 -0.774449  -0.230139
296  0.685181   0.303135
297 -0.105452   0.017481
298  0.472318  -0.191399
299  0.928453   0.172745
300 -0.409542  -0.506155
301 -1.048130  -0.903656
302 -0.652813  -0.227237
303  0.411500   0.783856
304 -0.835267  -0.586179
305  0.381091   1.325862
306 -0.713631  -0.579530
307 -0.804858  -0.482637
308  1.019680   0.441848
309 -0.196679   0.117203
310  1.810313   1.754815
311 -0.196679   0.733378
312 -0.287906  -0.753661
313 -0.987312  -0.856852
314  0.624363   0.842676
315 -0.622404  -0.788178
316 -1.200175  -0.834473
317 -0.865676  -0.370189
318 -0.987312  -0.765268
319 -1.139357  -1.003652

[320 rows x 2 columns]

In [183]:
from sklearn import metrics  
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred))  
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))  
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))  

Mean Absolute Error: 0.4408379007098189
Mean Squared Error: 0.36037896897724037
Root Mean Squared Error: 0.6003157244127796


# Result:

Wheat Seed Logistic regression:<br>
- Normal
0.9285714285714286<br>
- PCA
0.9047619047619048<br>
- KPCA
0.6904761904761905<br>


Wine Data Linear regression:<br>
- Normal
Mean Absolute Error: 15.893100141029436<br>
Mean Squared Error: 560.1825876096477<br>
Root Mean Squared Error: 23.668176685364838 <br>

- PCA    
Mean Absolute Error: 0.4408379007098189<br>
Mean Squared Error: 0.36037896897724037<br>
Root Mean Squared Error: 0.6003157244127796<br>
- KPCA
Mean Absolute Error: 2584015151647.59<br>
Mean Squared Error: 1.4246020099333066e+25<br>
Root Mean Squared Error: 3774390030101.959<br>

